src/crimson/osd/osd_operations/pg_splitting.cc

```diff
@@ -0,0 +1,150 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <seastar/core/future.hh>
+
+#include "include/types.h"
+#include "common/Formatter.h"
+#include "crimson/osd/pg.h"
+#include "crimson/osd/osdmap_service.h"
+#include "crimson/osd/shard_services.h"
+#include "crimson/osd/osd_operations/pg_advance_map.h"
+#include "crimson/osd/osd_operations/pg_splitting.h"
+#include "crimson/osd/osd_operation_external_tracking.h"
+#include "osd/PeeringState.h"
+
+namespace {
+  seastar::logger& logger() {
+    return crimson::get_logger(ceph_subsys_osd);
+  }
+}
+
+namespace crimson::osd {
+
+PGSplitting::PGSplitting(
+  Ref<PG> pg, ShardServices &shard_services, OSDMapRef cur_map, epoch_t next_epoch,
+  PeeringCtx &&rctx)
+  : pg(pg), shard_services(shard_services), cur_map(cur_map), next_epoch(next_epoch),
+    rctx(std::move(rctx))
+{
+  logger().debug("{}: initiated", *this);
+}
+
+PGSplitting::~PGSplitting() {}
+
+void PGSplitting::print(std::ostream &lhs) const
+{
+  lhs << "PGSplitting("
+      << "pg=" << pg->get_pgid()
+      << " from=" << cur_map->get_epoch()
+      << " to=" << next_epoch;
+  lhs << ")";
+}
+
+void PGSplitting::dump_detail(Formatter *f) const
+{
+    f->open_object_section("PGSplitting");
+    f->dump_stream("pgid") << pg->get_pgid();
+    f->dump_int("from", cur_map->get_epoch());
+    f->dump_int("to", next_epoch);
+    f->close_section();
+}
+
+
+seastar::future<> PGSplitting::start()
+{
+  logger().debug("{}: start", *this);
+  unsigned old_pg_num;
+  using cached_map_t = OSDMapService::cached_map_t;
+
+  if (cur_map->have_pg_pool(pg->get_pgid().pool())) {
+    old_pg_num = cur_map->get_pg_num(pg->get_pgid().pool());
+  } else {
+    logger().debug("{} pool doesn't exist in {}", pg->get_pgid(), cur_map);
+    return seastar::now();
+  }
+  
+  return shard_services.get_map(next_epoch).then(
```

Comment:
If we will pass the actual map to `PGSplitting` without trying to obtain it here, we could calculate any splits from within the PGSplitting ctor as this won't require any futureized calls (because we have all the information needded).
That way, `children`, `split_pgs` and so on could be members of `PGSplitting` which would be set before `start` invocation and we won't need to keep them alive with a dedicated `do_with`.